[LLDB] Add type summaries for MSVC STL strings

Nerixyz opened the pull request:
This PR adds type summaries for `std::{string,wstring,u8string,u16string,u32string}` from the MSVC STL.

See https://github.com/llvm/llvm-project/issues/24834 for the MSVC STL issue.

The following changes were made:

- `dotest.py` now detects if the MSVC STL is available. It does so by looking at the target triple, which is an additional argument passed from Lit. It specifically checks for `windows-msvc` to not match on `windows-gnu` (i.e. MinGW/Cygwin).
- (The main part): Added support for summarizing `std::(w)string` from MSVC's STL. Because the type names from the libstdc++ (pre C++ 11) string types are the same as on MSVC's STL, `CXXCompositeSummaryFormat` is used with two entries, one for MSVC's STL and one for libstdc++.
  With MSVC's STL, `std::u{8,16,32}string` is also handled. These aren't handled for libstdc++, so I put them in `LoadMsvcStlFormatters`.

Nerixyz commented:
I would love some general feedback on this PR. I updated the description to reflect the current state.

For some reason, the Windows CI doesn't run on this PR, but I ran the STL tests locally.

DavidSpickett commented:
> Added a new type summary kind CXXCompositeSummaryFormat. It holds a list of child summaries paired with a validator function that checks if a ValueObject can be formatted by a child. This feels like it should be done in another PR, but there aren't any tests for it other than the STL formatters (since it's only in C++ it can't be used in Python).

I had a similar problem with another formatting option, so I added https://github.com/llvm/llvm-project/blob/main/lldb/unittests/ValueObject/DumpValueObjectOptionsTests.cpp.

Maybe you could extend or take inspiration from that.

Michael137 reviewed (commented):
Thanks for working on this! The general idea of having composite formats seems reasonable to me (though others, CC @jimingham, might have opinions). If libc++ ever decided to majorly overhaul some types we might be able to provide backwards compatibility for old layouts this way. Will give it a more thorough review later though.

labath commented:
The composite formats are not an *un*reasonable idea, but I can't help but wonder if we even need to touch the generic data formatter infrastructure. I mean we already have the ability to do callback-based summaries, so what if we just did the stl switch inside the callback? Being able to do this declaratively is sort of nice, but this already isn't completely declarative since you have callbacks for type validation.

Maybe this means that we can't specify the summary for std::string via summary strings (or maybe we can -- I haven't checked how the code is structured), but I don't think that's such a disaster. Implementing `${var._M_dataplus._M_p}` in c++ is not that hard, and this is probably one of our more complex summaries (containers just say `size=#children`), which doesn't even need to be customized for different STLs.

Nerixyz committed:
[LLDB] Add type summaries for MSVC STL strings

Nerixyz commented:
> The composite formats are not an _un_reasonable idea, but I can't help but wonder if we even need to touch the generic data formatter infrastructure. I mean we already have the ability to do callback-based summaries, so what if we just did the stl switch inside the callback? Being able to do this declaratively is sort of nice, but this already isn't completely declarative since you have callbacks for type validation.

That's a good point. I removed the `CXXCompositeSummaryFormat` and implemented the selection in a lambda.

> Implementing `${var._M_dataplus._M_p}` in c++ is not that hard, and this is probably one of our more complex summaries (containers just say `size=#children`), which doesn't even need to be customized for different STLs.

Right, I did that at the beginning, but then saw more types using string summaries without looking at what they're doing, so I wanted to support both.

Without the new summary type this PR is much more manageable.

Nerixyz commented:
Another aside: Does anyone know why the [libstdc++ `std::__cxx11::string` summaries](https://github.com/llvm/llvm-project/blob/802fa92aee3565768887615108aa3e924d4e0fc7/lldb/source/Plugins/Language/CPlusPlus/LibStdcpp.cpp#L240-L358) weren't using the string summary? They were initially added in [D13964](https://reviews.llvm.org/D13964), but it doesn't provide a reason to do so. Note that the summary for a `std::string v = "foo"` was the format string one (`${var._M_dataplus._M_p}`). Only when printing a reference or pointer to a string, `LibStdcppStringSummaryProvider` was called. I tried to use this function for all libstdc++ strings, but this broke when using `-D_GLIBCXX_USE_CXX11_ABI=0` (on MinGW). On the other hand, `${var._M_dataplus._M_p}` works with both ABIs.

labath commented:
> Another aside: Does anyone know why the [libstdc++ `std::__cxx11::string` summaries](https://github.com/llvm/llvm-project/blob/802fa92aee3565768887615108aa3e924d4e0fc7/lldb/source/Plugins/Language/CPlusPlus/LibStdcpp.cpp#L240-L358) weren't using the string summary? They were initially added in [D13964](https://reviews.llvm.org/D13964), but it doesn't provide a reason to do so. Note that the summary for a `std::string v = "foo"` was the format string one (`${var._M_dataplus._M_p}`). Only when printing a reference or pointer to a string, `LibStdcppStringSummaryProvider` was called. I tried to use this function for all libstdc++ strings, but this broke when using `-D_GLIBCXX_USE_CXX11_ABI=0` (on MinGW). On the other hand, `${var._M_dataplus._M_p}` works with both ABIs.

Well, the commit says (I haven't verified this) that the std::string formatter at the time was written in python, so this could have been a way to avoid the python dependency. As for the implementation, a possible reason (again, I don't know if that's the case) could be to make the formatter work in the case where you don't have debug info for std::string -- it looks like it avoids looking at any type information, which means it could work even with a forward declaration -- but that it would break when the ABI changes, as you've noticed.

I don't think we need to concern ourselves with that though. Your new implementation is fine.

Nerixyz commented on lldb/packages/Python/lldbsuite/test/dotest.py:
```diff
@@ -299,6 +299,8 @@ def parseOptionsAndInitTestdirs():
     configuration.cmake_build_type = args.cmake_build_type.lower()
 
+    configuration.target_triple = args.target_triple
+
```
There's no `-stdlib`, but detection through `_MSVC_STL_VERSION` works.

labath replied:
That's better, though it would be even better if we could run these tests even if msvcstl was not the default clang target (similar to how we can run libc++ tests even when clang targets libstdc++ by default). 

From what I gather, this behavior is controlled by the target triple. Would manually changing the target triple by appending something to CFLAGS (take `self.dbg.GetSelectedPlatform().GetTriple()` and append "-msvc" or something) be a terrible idea? (FWIW, [this](https://github.com/llvm/llvm-project/blob/6dad1e87fb7a8de7033a315f5dd5d7c95dab32f6/lldb/packages/Python/lldbsuite/test/make/Makefile.rules#L371) is where we append libc++ flags.

labath replied:
> In the MSVC toolchain on Clang, [there's an open FIXME](https://github.com/llvm/llvm-project/blob/fcd82f9cf3b64d8959afed4569259163da3a5afd/clang/lib/Driver/ToolChains/MSVC.cpp#L786) for supporting libc++. I feel hesitant changing the target there.

Yeah, that's not what I'm suggesting.

> 
> If I understand you correctly, then the test would check if it's running on `*-pc-windows-*` and change it to `*-pc-windows-msvc`. In which case, it assumes to always run with MSVC's STL. That could work. Clang on MinGW uses `x86_64-w64-windows-gnu` so this wouldn't get changed, which is good.

Yes, something like that.

> Maybe it's better to get that fixme fixed in Clang (i.e. introduce something like `-stdlib=libc++`/`-stdlib=msvcstl`). As far as I know, MSVC's STL only works on the `*-msvc` targets.

Maybe -- I don't know. It may be there is a reason this isn't implemented that way. It may have something to do with the fact that the triple also controls other things (such as the name mangling). Maybe MSVC STL doesn't work with the itanium mangling (why should it)?

This is why I'm not totally convinced that this is a good idea, though I also believe that it's important that these tests do get run -- somewhere. If you can confirm that the tests would run on the existing windows bots (we can check with bot owners if needed), then maybe this wouldn't be necessary.

DavidSpickett replied:
I will check tomorrow whether with the current content of the PR, the tests would run on the bot.

DavidSpickett replied:
They do run:
```
UNSUPPORTED: LLDB (C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\clang.exe-aarch64) :: test_with_run_command_dsym (TestDataFormatterMsvcStlU8String.MsvcStlU8StringDataFormatterTestCase.test_with_run_command_dsym) (test case does not fall in any category of interest for this run)
PASS: LLDB (C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\clang.exe-aarch64) :: test_with_run_command_dwarf (TestDataFormatterMsvcStlU8String.MsvcStlU8StringDataFormatterTestCase.test_with_run_command_dwarf)
UNSUPPORTED: LLDB (C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\clang.exe-aarch64) :: test_with_run_command_dwo (TestDataFormatterMsvcStlU8String.MsvcStlU8StringDataFormatterTestCase.test_with_run_command_dwo) (test case does not fall in any category of interest for this run)
----------------------------------------------------------------------
Ran 3 tests in 2.359s

UNSUPPORTED: LLDB (C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\clang.exe-aarch64) :: test_with_run_command_dsym (TestDataFormatterMsvcStlString.MsvcStlStringDataFormatterTestCase.test_with_run_command_dsym) (test case does not fall in any category of interest for this run)
PASS: LLDB (C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\clang.exe-aarch64) :: test_with_run_command_dwarf (TestDataFormatterMsvcStlString.MsvcStlStringDataFormatterTestCase.test_with_run_command_dwarf)
UNSUPPORTED: LLDB (C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\build\bin\clang.exe-aarch64) :: test_with_run_command_dwo (TestDataFormatterMsvcStlString.MsvcStlStringDataFormatterTestCase.test_with_run_command_dwo) (test case does not fall in any category of interest for this run)
----------------------------------------------------------------------
Ran 3 tests in 1.936s

OK (skipped=2)
```
You should add the confusingly named `NO_DEBUG_INFO_TESTCASE = True` to the tests so that we don't try to run the dsym and dwo forms.

There are also some unexpected passes, which is probably a good thing:
```
  lldb-api :: commands/expression/call-function/TestCallStdStringFunction.py
  lldb-api :: functionalities/data-formatter/data-formatter-skip-summary/TestDataFormatterSkipSummary.py
  lldb-api :: functionalities/data-formatter/stringprinter/TestStringPrinter.py
  lldb-api :: python_api/sbvalue_persist/TestSBValuePersist.py
```
I did apply your commit onto main, rather than building the PR branch so it could be a different change causing that.

See if you have those locally. Ideally they are unexpected passes everywhere now.

DavidSpickett replied:
Also I'm bothered that there isn't any explicit enabling of the MSVC STL in the compiler flags, but I suppose that's what you and Pavel have been debating this whole time :)

What you've got is a predicate that says "does this compiler by default use MSVC STL" then the tests deliberately don't add any stlib related flags, so that they get that default. And as you mentioned above, there is no "opt in" to MSVC STL anyway.

Do I understand correctly? If so I think that is a reasonable way to do this. The fact that it relies on implicit selection seems to be out of your control.

labath replied:
Yeah, I think it's fine. FWIW, we also don't add any special flags when the compiler is gcc and we're running libstdc++ tests, only that fact is obscured by a lot of ifdefs.

labath commented on lldb/source/Plugins/Language/CPlusPlus/LibStdcpp.cpp:
```diff
@@ -239,122 +240,24 @@ VectorIteratorSyntheticFrontEnd::GetIndexOfChildWithName(ConstString name) {
-    }
-  }
-  return false;
+  ValueObjectSP dataplus = valobj.GetChildMemberWithName("_M_dataplus");
```
Could this be implemented the same way as `LibStdcppStringSummaryProvider` (maybe even by the same function)?

labath reviewed (commented):
I think the tests are fine. @Michael137, @jimingham, do you have anything to say about the implementation?

Michael137 commented on lldb/include/lldb/DataFormatters/StringPrinter.h:
```diff
@@ -152,6 +152,20 @@ class StringPrinter {
   ReadBufferAndDumpToStream(const ReadBufferAndDumpToStreamOptions &options);
+
+  template <StringElementType element_type>
+  static constexpr uint64_t ElementByteSize() {
```
Can we just make this a static in `MsvcStl.cpp`?

Michael137 commented on lldb/source/Plugins/Language/CPlusPlus/CPlusPlusLanguage.cpp:
```diff
@@ -1372,6 +1373,37 @@ static void LoadLibCxxFormatters(lldb::TypeCategoryImplSP cpp_category_sp) {
+  category_sp->AddTypeSummary(
+      makeSpecifier(llvm::formatv("std::basic_string<{}>", char_ty).str()),
+      summary_sp);
+  // std::basic_string<char,std::char_traits<char>,std::allocator<char> >
+  category_sp->AddTypeSummary(
+      makeSpecifier(llvm::formatv("std::basic_string<{0},std::char_traits<{0}>,"
+                                  "std::allocator<{0}> >",
+                                  char_ty)
+                        .str()),
+      summary_sp);
```
Out of curiosity, what tool generates the demangled string in this way?

Michael137 commented on lldb/source/Plugins/Language/CPlusPlus/LibStdcpp.cpp:
```diff
@@ -239,122 +240,24 @@ VectorIteratorSyntheticFrontEnd::GetIndexOfChildWithName(ConstString name) {
+  if (!ptr)
+    return false;
+
+  stream << ptr->GetSummaryAsCString();
```
Is this a simplification that we can only do with this patch? Or can this be a separate PR?

Nerixyz replied:
Opened https://github.com/llvm/llvm-project/pull/146562.

Michael137 commented on lldb/source/Plugins/Language/CPlusPlus/LibStdcpp.h:
```diff
@@ -18,11 +18,11 @@ namespace lldb_private {
 bool LibStdcppStringSummaryProvider(
     ValueObject &valobj, Stream &stream,
-    const TypeSummaryOptions &options); // libcstdc++ c++11 std::string
+    const TypeSummaryOptions &options); // libcstdc++ std::string
```
```suggestion
    const TypeSummaryOptions &options); // libstdc++ std::string
```

Michael137 commented on lldb/source/Plugins/Language/CPlusPlus/LibStdcpp.h:
```diff
@@ -18,11 +18,11 @@ namespace lldb_private {
 bool LibStdcppWStringSummaryProvider(
     ValueObject &valobj, Stream &stream,
-    const TypeSummaryOptions &options); // libcstdc++ c++11 std::wstring
+    const TypeSummaryOptions &options); // libcstdc++ std::wstring
```
```suggestion
    const TypeSummaryOptions &options); // libstdc++ std::wstring
```

Michael137 commented on lldb/test/API/functionalities/data-formatter/data-formatter-stl/msvcstl/string/TestDataFormatterMsvcStlString.py:
```diff
@@ -0,0 +1,120 @@
+# coding=utf8
+"""
+Test std::*string summaries with MSVC's STL.
+"""
```
I might've missed some discussion on this:

Not a blocker for this PR but why do we have separate API test directories for the various STLs? The only difference in them being `USE_LIBCPP` vs. `USE_LIBSTCPP` in the Makefiles. And now a third one with no apparent difference in the Makefile.

We should probably consolidate them all into the `data-formatter/data-formatter-stl` and just pass the necessary makefile flags from within the python tests (via `self.build`)?

@labath @DavidSpickett anything stopping us from doing that?

labath replied:
I think it's mainly historical. We have `data-formatter/data-formatter-stl/generic/`, but it's a fairly recent addition. I think it would be nice to move everything towards that, in particular because it encourages a consistent presentation of different implementations -- not just for humans, but this also makes it easier to write data formatters for other data structures (which contain std:: types internally) -- it's nice when the data formatter works regardless of the STL version your code is compiled with.

In theory there could be tests that only make sense on a particular (e.g. a test for what happens when a particular internal pointer is corrupted), but those could be handled with ifdefs, or with separate tests.

Michael137 reviewed (commented):
LGTM, with some minor nits

Michael137 reviewed (approved):
LGTM once https://github.com/llvm/llvm-project/pull/143177#discussion_r2177093572 is merged separately.

Michael137 commented:
@labath @jimingham any final comments? Otherwise happy to merge
